src/functional/ffront/single_static_assign.py

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import ast
+from dataclasses import dataclass, field
+
+
+class SingleStaticAssignPass(ast.NodeTransformer):
+    class RhsRenamer(ast.NodeTransformer):
+        """
+        Rename right hand side names.
+
+        Only read from parent visitor state, can not modify.
+        """
+
+        def __init__(self, state):
+            super().__init__()
+            self.state = SingleStaticAssignPass.State(name_counter=state.name_counter.copy())
```

Comment:
I don't see the point of cloning the state here if it's never going to be modified. If you do it just for protection, you could define `name_counter` in `State` as a generic `Mapping` and here create a new `State` using `MappingProxyType` as a wrapper around the input state, which cannot be update it (but as I said, it might be a bit of overkill)

Reply:
I will just pass the state through. I agree that making mutation impossible is over-engineering.